### ISSUE TITLE
Can't receive alerts on pageduty from wazuh

### ISSUE THREAD

**Issue by rishit794** [wazuh/wazuh#15457]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| X.Y.Z-rev | Wazuh component | Manager/Agent | Packages/Sources | OS version |

**Comment by rishit794** [comment]
Hi Everybody, I am new to wazuh and have an all-in-one wazuh setup. I have added the pagerduty module into the ossec.conf file... It works well. However, I see only one type of alert on the pagerduty dashboard which is "Wazuh Alert: 'Audit: Watch - Read access........" any idea why I am not able to see other alerts?  Just for the testing purpose, I created a file on one of the agent VM and I see that the alert is appearing under the integrity monitroing tab on the wazuh dashboard but cannot be found in the alerts.log file and also cannot be seen on pagerduty along with other alert... I have added the correct intergration block in ossec.conf file and also added the correct syscheck config in ossec.conf file of both the agent and manager. 
Any help would be greatly appreciated

**Comment by rishit794** [comment]
Hi Carlos,

First of all I really appreciate that you're helping me with this.  Below
is the required information


   1. The Wazuh version of the manager and the agent. =  4.3.8
   2. Share with us your Pagerduty configuration in the ossec.conf. Remove
   sensitive information such as the api key.

              <integration>
                 <name>pagerduty</name>
                 <api_key></api_key> 
             </integration>

          3.Confirm us that you have configured Pagerduty in the manager
and not in the agent = Yes I confirm that I have configured pagerduty only
in manager

As per your suggestion, I have also added integrator.debug=2 into the
/var/ossec/etc/local_internal_options.conf file in manager. Let me know if
you require any further information.


*Regards,*
*Rishit Sheth*

**Comment by CarlosRS9** [comment]
Your configuration seems valid and your Wazuh installation contains the latest changes to the integrator module, so that shouldn't be the problem either.

Have you checked the `/var/ossec/logs/integrator.log` to see if there are any error messages or anything that looks strange after restarting the manager and forcing the generation of an alert again? Also, have you been able to check if all your alerts are finally appearing on the dashboard?

**Comment by rishit794** [comment]
Hi Carlos,

Yes I checked in the /var/ossec/logs/integrator.log file but cannot see any
errors. and also all modules on the wazuh dashboard have their own events
showing. However, I still see only maximum 4 alerts on pagerduty and most
of the time they are related to Audit watch access. Any idea about this?

**Comment by rishit794** [comment]
HI Carlos,

I want to be a little clear about my previous email. for an example, I want
something where when I create a file on one of the agent VMs , it should be
logged into alerts.log & alerts.json file. and should be visible under
"file integrity monitoring" tab of wazuh dashboard and also route that
alert to pagerduty and I should be notified so I can go and check
whether the addition of file is legit or not.

So I tried doing something like this.. I tried creating & deleting a file
on an agent vm under my home directory which is "/home/rishit" ... I saw
that the alert appeared on the "events" tab of a file integrity monitoring
on wazuh dashboard.  However, I cannot see that alert in alerts.log &
alerts.json file on manager. As I understood, if the alert is showing on
wazuh dashboard, then it should be there in alerts.log & alerts.json file,
which is not happening in my case and also that alert is not being reported
on pagerduty.

So, basically this is the issue which I want to figure out and want to see
an alert on pagerduty as soon as a file is added/deleted on wazuh agent (of
course I have added directory which needs to be monitored in to the
<syscheck> of agent)

I have attached a screenshot for your reference. Let me know if you require
any further information.

[image: Screen Shot 2022-11-23 at 3.41.35 PM.png]

**Comment by CarlosRS9** [comment]
> HI Carlos, I want to be a little clear about my previous email. for an example, I want something where when I create a file on one of the agent VMs , it should be logged into alerts.log & alerts.json file. and should be visible under "file integrity monitoring" tab of wazuh dashboard and also route that alert to pagerduty and I should be notified so I can go and check whether the addition of file is legit or not. 

This is exactly how it is supposed to work by default. It's strange because the alerts file is the intermediate step between the analysis engine and the dashboard. When an event generates an alert, it is reflected in the alerts and then sent to the dashboard.

You are using an all-in-one installation. By default these installations do not make use of worker nodes. That is, you don't have a cluster and there is only one manager in your environment. Confirm me that this is your case, because otherwise you might be looking at the logs on the worker node instead of the master, as both are managers, which would explain the alerts log file issue.


Anyway, if you don't mind let's do the following proof of concept. All these steps must be done in the manager unless you are told otherwise. I'm going to explain it step by step so that any other users who finds themselves in a similar situation can follow it too.


First make sure that you have enabled the following option under `<global>` in your `ossec.conf`: 
```
<alerts_log>yes</alerts_log>
```

Then configure the file monitoring using the syscheck module. This can be applied in the manager or in an agent. For example:
```
<directories realtime="yes">/test</directories>
```

Please, do note that file monitoring can be configured in 3 different ways: `realtime`, `whodata` and `scheduled`. If you don't indicate `realtime="yes"` or `whodata="yes"` your alerts will only be generated when the time indicated in `frequency` is reached, as `scheduled` is the default behaviour. For our test we are going to use the `realtime` mode.

Once configured, created the directory that we are going to monitor in the manager or agent. In this case:
```
mkdir /test
```

Now enable integrator debug mode by adding the following line to the `/var/ossec/etc/local_internal_options.conf` file in the manager:
```
integrator.debug=2
```

Before restarting the manager to apply the configuration, I recommend opening several terminals or tabs to monitor each of the following files in real time:
```
tail -f /var/ossec/logs/ossec.log
```
```
tail -f /var/ossec/logs/integrations.log
```
```
tail -f /var/ossec/logs/alerts/alerts.log
```

Once everything is ready, restart the manager.

After restarting the manager if you create a new file in the monitored directory you should see the following:

- The expected output in the `ossec.log` should look like this:
```
2022/11/24 10:48:03 wazuh-integratord[2996] integrator.c:133 at OS_IntegratorD(): DEBUG: sending new alert.
2022/11/24 10:48:03 wazuh-integratord[2996] integrator.c:381 at OS_IntegratorD(): DEBUG: file /tmp/pagerduty-1669286883-155329501.alert was written.
2022/11/24 10:48:03 wazuh-integratord[2996] integrator.c:394 at OS_IntegratorD(): DEBUG: Running: integrations /tmp/pagerduty-1669286883-155329501.alert API-KEY  debug
2022/11/24 10:48:03 wazuh-integratord[2996] integrator.c:403 at OS_IntegratorD(): DEBUG: integratord:   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
2022/11/24 10:48:03 wazuh-integratord[2996] integrator.c:403 at OS_IntegratorD(): DEBUG: integratord:                                  Dload  Upload   Total   Spent    Left  Speed
100   520  100    19  100   501     22    602 --:--:-- --:--:-- --:--:--   624(): DEBUG: integratord: 
2022/11/24 10:48:04 wazuh-integratord[2996] integrator.c:416 at OS_IntegratorD(): DEBUG: Command ran successfully.

```
- The expected output in the `integrations.log` should look like this:
```
Thu Nov 24 10:46:43 UTC 2022 integrations/pagerduty /tmp/pagerduty-1669286803--1259645184.alert YOUR-API-KEY
```

- And the expected output for the `alerts.log` should look like this:
```
** Alert 1669287951.7358: - ossec,syscheck,syscheck_entry_added,syscheck_file,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,tsc_PI1.4,tsc_PI1.5,tsc_CC6.1,tsc_CC6.8,tsc_CC7.2,tsc_CC7.3,
2022 Nov 24 11:05:51 wazuh->syscheck
Rule: 554 (level 5) -> 'File added to the system.'
File '/test/testfile' added
Mode: realtime

Attributes:
 - Size: 0
 - Permissions: rw-r--r--
 - Date: Thu Nov 24 11:05:51 2022
 - Inode: 31491
 - User: root (0)
 - Group: root (0)
 - MD5: d41d8cd98f00b204e9800998ecf8427e
 - SHA1: da39a3ee5e6b4b0d3255bfef95601890afd80709
 - SHA256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

Please follow all these steps, check that the outputs you get are similar, that there are no errors or messages indicating something is wrong and that finally the alert appears on the dashboard.


I look forward to hearing from you after performing this test so I can continue to help you.

**Comment by rishit794** [comment]
Thanks a lot Carlos for detailed information.
I performed following steps:

1. Under the <global> section of ossec.conf file I have
<alerts_log>yes</alerts_log>
2. In the manager's ossec.conf file I added <directories
realtime="yes">/test</directories>
3. Created /test directory
4. I checked that I have integrator.debug=2 in
/var/ossec/etc/local_internal_options.conf
5. opened couple of sessions to monitor continuous logs in following
locations
    tail -f /var/ossec/logs/ossec.log

   tail -f /var/ossec/logs/integrations.log

   tail -f /var/ossec/logs/alerts/alerts.log

6. Restarted the manager service and created a file testing.txt under
location /test
7. As soon as I created a test file I was able to capture alerts into
integrations.log file and ossec.log file. However, I cannot see any
event/alert related to "/test/testing.txt file was created"into alerts.log
& alerts.json file... Any idea?

**Comment by CarlosRS9** [comment]
It seems there is indeed an issue in your environment related to File Integrity Monitoring. Can you confirm that you are not performing any operation with the alerts files such as migration, rotation, deletion or similar?

On the other hand, can you share with us exactly what steps you have followed to install Wazuh? If you used a particular link from our documentation, please share it here. Then I can report this to the FIM team so they can check for any errors in the module.

Sorry for the inconvenience.

**Comment by rishit794** [comment]
Hi Carlos,

Not sure how but now when I create a file I can see an alert in alerts.log
file. Apart from this, I added <rule_id>80790</rule_id> under the pagerduty
section and restarted the manager. After this I was able to see the alert
on pagerduty as soon as I created a file on any agent server. So basically
In my opinion when I specifically added rule id 80790 it replaced the
default id 80784 which performs only read access. So, I think I am good now.

However, I have a couple of questions. For now, almost all the alerts are
going from wazuh to pagerduty, Is it possible to limit the number of alerts
which are going from wazuh to pagerduty?  Can you explain or provide a
document which describes how to create a rule-group?

**Comment by CarlosRS9** [comment]
I'm glad to see that it's working properly.

> Is it possible to limit the number of alerts which are going from wazuh to pagerduty?

It is possible to specify which alert IDs are sent to Pagerduty by providing a list of accepted rule IDs as you have done. It is also possible to specify which rule groups should be taken into account using the [group](https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/integration.html#group) tag. 

If what you mean is that you want to limit the number of alerts for a given rule that are sent to PagerDuty, I'm afraid that is not possible to do through the PagerDuty configuration. On the other hand, it is possible to create your own custom rules that are triggered when an alert occurs X times in a row and then send that specific alert to PagerDuty.

> Can you explain or provide a document which describes how to create a rule-group?

Rule groups are very easy to use. If you look at the existing rules you will see that they are stored inside a `group` tag with a `name` attribute. For example, all audit rules are located in a group called `audit`. Creating your own rule groups is as easy as declaring them in your `/var/ossec/etc/rules/local_rules.xml`, following this structure:
```
<group name="local,syslog,sshd,">

  <rule id="100001" level="5">
    <if_sid>5716</if_sid>
    <srcip>1.1.1.1</srcip>
    <description>sshd: authentication failed from IP 1.1.1.1.</description>
    <group>authentication_failed,pci_dss_10.2.4,pci_dss_10.2.5,</group>
  </rule>

</group>
```
As you can see, there is a group declaration with one or more rules inside it. You can have as many groups as you want within your `local_rules.xml` and you can enter as many rules as you want within each of your groups. Remember that the ID of each custom rule must be unique.

Groups can have multiple values for its name attribute, as long as they are separated by `,` and without empty spaces. Here are some valid examples:
```
<group name="local,syslog,sshd,"></group>
<group name="pagerduty,custom_group"></group>
<group name="login,"></group>
```
Remember that the names should always end with a comma.

Finally, if you want to learn more about how the rules work and how to create your own I suggest you take a look at the following links to learn more about the rules

- [Change the rules](https://documentation.wazuh.com/current/learning-wazuh/replace-stock-rule.html?highlight=local_rules)
- [Custom rules and decoders](https://documentation.wazuh.com/current/user-manual/ruleset/custom.html)
- [Rules Syntax](https://documentation.wazuh.com/current/user-manual/ruleset/ruleset-xml-syntax/rules.html)

I hope you find this information useful. Let me know if you have any further questions.

**Comment by rishit794** [comment]
Thanks for the information Carlos. You have been very helpful. Just a side
question, Is there a document or article which explains what are different
types of rule ids used in wazuh and what exactly they do? I want to know
about it so I can add rule_id accordingly into the pagerduty configuration.

*Regards,*
*Rishit Sheth*

**Comment by rishit794** [comment]
Hi Carlos,

To be very specific on my previous question, when I go into management --->
Rules section on wazuh dashboard as shown in attached screenshot, I see in
total 4103 rules available and each of them are used for specific
monitoring. I don't want to receive alerts for all of them but If I have to
pick any particular rule_ids from this list and specify them into the
<rule_id></rule_id> section of <integration></integration> module
separated by comma, then what is the quickest way to do it? Do I need to
manually pick them one by one and then add them into <rule_id></rule_id> ?
Is there a possible faster way?

I want to receive alerts from only specific rule_ids for the events link
addition/deletion of a file, unauthorized access, authorized attempt of
login etc.. I would really appreciate it if you can help me to
understand how can I make it work.

**Comment by CarlosRS9** [comment]
Hi @rishit794 

There is no document explaining the different rule IDs available. However, you can check our ruleset in the [repository](https://github.com/wazuh/wazuh/tree/4.3/ruleset/rules) or using the UI. As you can see, the rules are stored in different `.xml` files. In those file you can find the rules you are interested. For example the `audit` rules, which includes file manipulation, are in `0365-auditd_rules`, while the `syslog` rules, which includes sudo operations, are in `0020-syslog_rules.xml`.

Regarding how to indicate which alerts you want to send to your PagerDuty integration, there are 3 ways to fine tune this:

1. You can use the `rule_id` as you commented, manually picking them one by one and adding them to your PagerDuty configuration.
1. You can use the `group` tag to send to Pagerduty only those alerts corresponding to the groups you are interested in. For example, you could have only alerts from `audit`, `syslog` and a group corresponding to your custom rules to be taken into account, if you wish.
3. You can have only those alerts sent whose alert level is high enough using the `level` tag. You can check the different alert levels and their descriptions in [this list](https://documentation.wazuh.com/current/user-manual/ruleset/rules-classification.html). Remember that you can modify the alert level of a default rule by creating a child rule that inherits from it and increasing its alert level there. 

Finally, take into account that you can combine these 3 options to fine tune the behaviour you want to obtain.

**Comment by CarlosRS9** [comment]
I see you have already been answered in that other thread @rishit794.

I'm closing this issue as it looks like it has been resolved. Please feel free to reopen this issue if you continue to have the problem with PagerDuty that we were discussing.
